### PR TITLE
Autopinning

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

File: Sources/Commands/PinsStore.swift [hunk 1]
```diff
@@ -90,12 +90,14 @@ public struct PinsStore {
...
     /// Unpin a pinnned repository and saves the state.
     ///
+    /// - Precondition: Autopin should be off.
     /// - Parameters:
     ///   - package: The package name to unpin. It should already be pinned.
     /// - Returns: The pin which was removed.
     /// - Throws: PinOperationError
     @discardableResult
     public mutating func unpin(package: String) throws -> Pin {
+        precondition(!autoPin, "Trying to unpin individual package with autopin on.")
```
Comment (ddunbar) [inline Sources/Commands/PinsStore.swift]: Why not throw here, versus precondition? Client is forced to handle in either case, but throw is a little more obvious.

File: Sources/Commands/Workspace.swift [hunk 2]
```diff
@@ -582,20 +582,30 @@ public class Workspace {
         try updateCheckouts(with: updateResults)
         // If we're repinning, update the pins store.
         if repin {
-            // Create a dictionary of result for fast lookup.
```
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: What should update do if autopin is on and repin isn't specified?

File: Sources/Commands/Workspace.swift [hunk 3]
```diff
@@ -803,6 +813,12 @@ public class Workspace {
             }
         }
 
+        // If autopin is enabled, unpin and pin everything.
+        if pinsStore.autoPin {
+            try pinsStore.unpinAll()
+            try pinAll()
```
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: Why do we have to unpin & then pinAll, vs just pinAll?

File: Sources/Commands/Workspace.swift [hunk 4]
```diff
@@ -816,12 +832,6 @@ public class Workspace {
         // Inform the delegate.
         delegate.removing(repository: dependency.repository.url)
 
-        // If this specifier is pinned, emit a note about stray pin.
-        if let pin = pinsStore.pins.first(where: { $0.repository == specifier }) {
-            // FIXME: Use diagnosics engine when we have that.
-            print("note: Considering unpinning \(pin.package), it is pinned at \(pin.version) but the package is being removed.")
-        }
-
```
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: Why does this go away? It still can apply with autopin off?

Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: This is moved above in `repinPackages(with:)` method where it is handled better. Now swiftpm will always emit this when repinning, having it here means it only emits once when repository is being removed.

File: Sources/Commands/Workspace.swift (hunk 3 above)
Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: pinAll pins all current manifest but doesn't remove old pins i.e stray pins.

File: Sources/Commands/Workspace.swift (hunk 2 above)
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: I'm tempted to say update with autopin on should be same as repin, I think it is what people will want (by default).

File: Sources/Commands/Workspace.swift (hunk 4 above)
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: Ah, tx!

File: Sources/Commands/Workspace.swift (hunk 3 above)
Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: These method could take a param `writeState: false`

Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: Or pinAll could take a "reset" option?

File: Sources/Commands/Workspace.swift (hunk 2 above)
Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: I will put that at tool level, seems more appropriate

**Pull request merged**